cmd/bcpd/app/app_test.go

```diff
@@ -457,12 +542,125 @@ func makeSendTx(t require.TestingT, chainID string, sender, receiver *account, t
+		Amount: &x.Coin{
+			Whole:  amount,
+			Ticker: ticker,
+		},
+		Memo: memo,
+	}
+
+	tx := &Tx{
+		Sum:      &Tx_SendMsg{msg},
+		Multisig: [][]byte{sender.id},
+	}
+
+	mandatorySigners := sender.signers()
+	for _, acc := range mandatorySigners {
+		sig, err := sigs.SignTx(acc.pk, tx, chainID, acc.nonce())
+		require.NoError(t, err)
+		tx.Signatures = append(tx.Signatures, sig)
+	}
+
+	txBytes, err := tx.Marshal()
+	require.NoError(t, err)
+	require.NotEmpty(t, txBytes)
+	return txBytes
+}
+
+func makeCreateContractTx(t require.TestingT, chainID string, signers [][]byte, threshold int64) *Tx {
+	msg := &multisig.CreateContractMsg{
+		Sigs:                signers,
+		ActivationThreshold: threshold,
+		AdminThreshold:      threshold,
+	}
+
+	return &Tx{
+		Sum: &Tx_CreateContractMsg{msg},
+	}
+}
+
+func newBlock(t require.TestingT, baseApp app.BaseApp, txs [][]byte, nbAccounts, blockSize, maxBlock, idx, height int) {
+	idTx := func(a, b int) int { return (a * blockSize) + b }
+	for k := 0; k < blockSize && idTx(idx, k) < maxBlock; k++ {
+		chres := baseApp.CheckTx(txs[idTx(idx, k)])
+		require.Equal(t, uint32(0), chres.Code, chres.Log)
+	}
+
+	header := abci.Header{Height: int64(height)}
+	baseApp.BeginBlock(abci.RequestBeginBlock{Header: header})
+
+	for k := 0; k < blockSize && idTx(idx, k) < maxBlock; k++ {
+		dres := baseApp.DeliverTx(txs[idTx(idx, k)])
+		require.Equal(t, uint32(0), dres.Code, dres.Log)
+	}
+
+	baseApp.EndBlock(abci.RequestEndBlock{})
+	cres := baseApp.Commit()
+	assert.NotEmpty(t, cres.Data)
+}
+
+// benchmarkSendTxWithMultisig runs the actual benchmark sequence with multisig eg.
+// N * CheckTx
+// BeginBlock
+// N * DeliverTx
+// EndBlock
+// Commit
+func benchmarkSendTxWithMultisig(b *testing.B, nbAccounts, blockSize, nbContracts, nbMultisigSigs int, threshold int64) {
+	id := func(i int64) []byte {
+		bz := make([]byte, 8)
+		binary.BigEndian.PutUint64(bz, uint64(i))
+		return bz
+	}
+
+	accounts := make([]*account, nbAccounts)
+	for i := 0; i < nbAccounts; i++ {
+		accounts[i] = &account{pk: crypto.GenPrivKeyEd25519()}
+	}
+
+	contracts := make([]*contract, nbContracts)
+	for i := 0; i < nbContracts; i++ {
+		sigs := make([]*account, nbMultisigSigs)
+		for k := 0; k < nbMultisigSigs; k++ {
+			sigs[k] = accounts[cmn.RandInt()%nbAccounts]
+		}
+		contracts[i] = &contract{id: id(int64(i + 1)), accountSigs: sigs, threshold: threshold}
+	}
+
+	chainID := "bench-net-22"
+	myApp := newMultisigTestApp(b, chainID, contracts)
+
+	for i, c := range contracts {
+		signer := accounts[cmn.RandInt()%nbAccounts]
+		c.id = createContract(b, myApp, chainID, int64(i+1), []*account{signer}, threshold, c.sigs()...)
+	}
+
+	txs := make([][]byte, b.N)
+	for i := 0; i < b.N; i++ {
+		sender := contracts[cmn.RandInt()%nbContracts]
+		recipient := contracts[cmn.RandInt()%nbContracts]
+		txs[i] = makeSendTxMultisig(b, chainID, sender, recipient, "ETH", "benchmark", 1)
+	}
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
```

Comment:
okay, i'm confused here, do we make b.N blocks, or b.N transactions?
i think this will call too many times and make a bunch of empty blocks at the end